---
title: "Heritability report"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output:
  html_document:
    code_folding: hide
    toc: true
---



The code chunk appears:
```{r, message = FALSE, warning = FALSE, echo = FALSE}
library(data.table)
library(tidyverse)
library(extrafont)
library(bootstrap)

# Declare constants
loadfonts()
old <- theme_set(theme_classic(base_size = 10))
theme_update(
  line = element_line(
    colour = "black", size = (0.5 / (ggplot2::.pt * 72.27/96)),
    linetype = 1, lineend = "butt", arrow = F, inherit.blank = T),
  strip.background = element_rect(colour = NA, fill = NA),
  panel.grid.minor = element_blank(),
  text = element_text(family="Helvetica"),
  title = element_text(colour = "#595A5C", face = "bold"),
)

expected_column_names <- c("p1", "p2", "rg", "se", "z", "p", "h2_obs", "h2_obs_se", "h2_int", "h2_int_se", "gcov_int", "gcov_int_se", "gene", "annot")

mr_genes <- fread("/gpfs/space/GI/eQTLGen/freeze2/eqtl_mapping/input/mr_GenesToInclude_2023-07-14.txt")
sampled_genes <- fread("/gpfs/space/GI/eQTLGen/freeze2/eqtl_mapping/input/sampled1K_and_mr_GenesToInclude_2023-10-05.txt")
```

```{r, message = FALSE, warning = FALSE, echo = FALSE}
gamma_moments_of_methods <- function(h2_obs, h2_obs_se) {
  median_squared_se <- median(h2_obs_se)^2
  print(median_squared_se)
  k_theta <- mean(h2_obs)
  print(k_theta)
  k_theta2 <- var(h2_obs_se) - median_squared_se
  print(k_theta2)

  theta <- k_theta / k_theta2
  print(theta)
  k <- k_theta / theta
  print(k)
  return(list("theta" = theta, "k" = k))
}

standard_errors_gw <- function(delete_values, mean_hsq) {
  mean_delete_values <- apply(delete_values, 1, mean)

  pseudovalues <- 200 * mean_hsq - 199 * mean_delete_values

  se <- sqrt(var(pseudovalues) / 200)
  return(se)
}

```

```{r}
median_exp_summary <- "/gpfs/space/GI/eQTLGen/freeze2/PreMetaQc/output/ExpMedianZSummary_2023-11-13.txt.gz"
exp <- fread(median_exp_summary) %>%
  pivot_longer(cols = -gene, names_to = "cohort", values_to = "median_exp") %>%
  group_by(gene) %>%
  summarise(median_exp = median(median_exp, na.rm=T))

gnomad_conservation <- fread("/gpfs/space/GI/eQTLGen/freeze2/Interpretation/heritability/data/gnomad.v4.0.constraint_metrics.tsv")

library(biomaRt)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

results <- getBM(attributes = c("ensembl_gene_id", "ensembl_transcript_id", "ensembl_peptide_id"),
                 filters = "ensembl_transcript_id", values = gnomad_conservation$transcript,
                 mart = mart)

gene_annotations <- gnomad_conservation %>% inner_join(results, by=c("transcript"="ensembl_transcript_id")) %>%
  inner_join(exp, by=c("ensembl_gene_id"="gene")) %>% distinct(ensembl_gene_id, .keep_all=TRUE)


```

```{r, message = FALSE, warning = FALSE, echo = FALSE}
input_path_cis <- "ldsc_table_cis.txt"
input_path_trans <- "ldsc_table_trans.txt"
input_path_gw <- "ldsc_table_gw.txt"

input_path_dt_trans <- "delete_values_combined_trans.tsv"
input_path_dt_gw <- "delete_values_combined_gw.tsv"

h2_table_cis <- fread(input_path_cis) %>%
  mutate(marker_gene = gene_id %in% mr_genes$ID)
h2_table_trans <- fread(input_path_trans) %>%
  mutate(marker_gene = gene_id %in% mr_genes$ID)
h2_table_gw <- fread(input_path_gw) %>%
  mutate(marker_gene = gene_id %in% mr_genes$ID)


delete_values_trans <- fread(input_path_dt_trans, data.table = F)
delete_values_gw <- fread(input_path_dt_gw, data.table = F)

h2_table_concat <- bind_rows(list("cis" = h2_table_cis, "trans" = h2_table_trans, "gw" = h2_table_gw), .id = "annot") %>%
  mutate(h2_z = h2_obs / h2_obs_se, h2_p = pnorm(q=h2_z, lower.tail=FALSE))

h2_dist <- h2_table_concat %>% filter(!marker_gene) %>% group_by(annot) %>%
  summarise(gamma = list(gamma_moments_of_methods(h2_obs, h2_obs_se))) %>%
  unnest_wider(gamma) %>%
  cross_join(tibble(h2_obs = seq(-0.001, 1, by=0.001))) %>%
  group_by(annot) %>%
  mutate(density = dgamma(h2_obs, scale=theta, shape=k))

h2_table_summary <- h2_table_concat %>% group_by(annot, marker_gene) %>%
  summarise(
      n_genes = length(unique(gene_id)),
      mean_h2 = mean(h2_obs),
      median_h2 = median(h2_obs),
      over_not = sum(h2_p < 0.05),
      h2_se = first(case_when(
          annot == "cis" ~ jackknife(h2_obs, mean)$jack.se,
          annot == "trans" ~ standard_errors_gw(delete_values_trans[,c(gene_id)], mean(h2_obs)),
          annot == "gw" ~ standard_errors_gw(delete_values_gw[,c(gene_id)], mean(h2_obs)))))

p <- ggplot(h2_table_concat %>% filter(!marker_gene), aes(h2_obs)) +
  geom_histogram(aes(y = after_stat(density))) +
  geom_line(data = h2_dist, aes(x=h2_obs, y=density), color = "#D55E00") +
  facet_wrap(vars(annot), scales="free_y", ncol = 2) +
  coord_cartesian(ylim = c(0, 10), xlim = c(NA, 1))

ggsave("distributions_h2_11-14_trans.pdf", p, width = 4, height = 4)

h2_table_concat_annot <- h2_table_concat %>% inner_join(exp, by = c("gene_id"="gene")) %>% filter(!marker_gene)

h2_table_pivot <- h2_table_concat_annot %>% filter(!marker_gene) %>% dplyr::select(gene_id, h2_obs, h2_obs_se, h2_z, h2_p, annot, median_exp) %>% pivot_wider(values_from = c("h2_obs", "h2_obs_se", "h2_z", "h2_p"), names_from = "annot")

p <- ggplot(h2_table_pivot, aes(h2_z_cis, h2_z_trans, color = log10(median_exp))) +
  geom_point(shape=16) + geom_smooth(method="lm")

ggsave("scatter_h2_11-21_trans.pdf", p, width = 4, height = 4)

p <- ggplot(h2_table_concat_annot, aes(log10(median_exp), h2_obs)) +
        geom_point(shape=16) +
        geom_smooth(method="lm") +
        facet_wrap(vars(annot), scales="free_y", ncol = 2)

ggsave("exp_h2_11-21_trans.pdf", p, width = 4, height = 4)

p <- ggplot(h2_table_concat_annot, aes(lof.pLI, h2_obs)) +
        geom_point(shape=16) +
        geom_smooth(method="lm") +
        facet_wrap(vars(annot), scales="free_y", ncol = 2)

ggsave("conservation_h2_11-14_trans.pdf", p, width = 4, height = 4)



h2_table_concat_sig <- h2_table_concat %>% inner_join(trans_signals, by = c("gene_id" = "Gene"))

p <- ggplot(h2_table_concat_sig, aes(n, h2_obs)) +
  geom_point() +
  geom_linerange(aes(ymin=h2_obs-h2_obs_se, ymax=h2_obs+h2_obs_se)) +
  geom_smooth() +
  facet_wrap(vars(annot), scales="free")

ggsave("distributions_h2_10-26_trans_sig.pdf", p, width = 6, height = 6)

```

```{r, message = FALSE, warning = FALSE, echo = FALSE}

mr_genes <- fread("/gpfs/space/GI/eQTLGen/freeze2/eqtl_mapping/input/mr_GenesToInclude_2023-07-14.txt")

res_table <- fread(input_path1, col.names=expected_column_names) %>%
    mutate(marker_gene = gene %in% mr_genes$ID)
res_table_noPCs <- fread(input_path2, col.names=expected_column_names) %>%
    mutate(marker_gene = gene %in% mr_genes$ID)

res_table_concat <- bind_rows(list("20PSs" = res_table, "NoPCs" = res_table_noPCs), .id = "meta_analysis")

h2_table_concat <- res_table_concat %>% filter(p1 == p2)

h2_table <- res_table %>% filter(p1 == p2)

h2_table %>% group_by(annot, marker_gene) %>%
    summarise(median_h2 = median(h2_obs),
              mean_h2 = mean(h2_obs))

trans_h2_table <- h2_table %>% filter(annot == "trans", !marker_gene)
cis_h2_table <- h2_table %>% filter(annot == "cis", !marker_gene)

h2_gamma_trans <- gamma_moments_of_methods(trans_h2_table$h2_obs, trans_h2_table$h2_obs_se)
h2_gamma_cis <- gamma_moments_of_methods(cis_h2_table$h2_obs, cis_h2_table$h2_obs_se)

draw <- tibble(x = seq(-0.001, 1, by=0.001)) %>%
    mutate(trans = dgamma(x, scale=h2_gamma_trans$theta, shape=h2_gamma_trans$k),
           cis = dgamma(x, scale=h2_gamma_cis$theta, shape=h2_gamma_cis$k)) %>%
    pivot_longer(cols = c(trans, cis), names_to = "annot", values_to = "h2_est")

p <- ggplot(h2_table, aes(rg)) +
    geom_histogram(aes(y = after_stat(density))) +
    #geom_line(data=draw, aes(x = x, y = h2_est), inherit.aes=F) +
    facet_wrap(vars(annot, marker_gene), scales="free")

ggsave("distributions_h2.pdf", p, width = 6, height = 6)

p <- ggplot(h2_table %>% filter(!marker_gene), aes(h2_obs, h2_obs_se)) +
    geom_point() +
    facet_wrap(vars(annot), scales="free")

ggsave("se_h2.pdf", p, width = 6, height = 6)


h2_table_concat %>% group_by(annot, marker_gene, meta_analysis) %>%
    summarise(median_h2 = median(h2_obs),
              mean_h2 = mean(h2_obs))


p <- ggplot(h2_table_concat, aes(h2_obs)) +
    geom_histogram(aes(y = after_stat(density))) +
    facet_wrap(vars(annot), scales="free")

ggsave("distributions_mr_h2.pdf", p, width = 6, height = 6)


p <- ggplot(h2_table %>% pivot_wider(id_cols = c(gene), values_from=c(h2_obs, h2_obs_se), names_from=annot),
            aes(h2_obs_cis, h2_obs_trans)) +
    geom_point() +
    geom_linerange(aes(xmin=h2_obs_cis-h2_obs_se_cis, xmax=h2_obs_cis+h2_obs_se_cis)) +
    geom_linerange(aes(ymin=h2_obs_trans-h2_obs_se_trans, ymax=h2_obs_trans+h2_obs_se_trans))

ggsave("scatter_h2.pdf", p)

p <- ggplot(h2_table,
            aes(x=gene, h2_obs)) +
    geom_point() +
    geom_linerange(aes(ymin=h2_obs-h2_obs_se, ymax=h2_obs+h2_obs_se)) +
    facet_wrap(vars(annot), scales="free_x") +
    theme(axis.text = element_text(angle = 90))

ggsave("range_h2.pdf", p)

combiend <- res_table_concat %>%
    inner_join((h2_table_concat %>% filter(annot == "trans") %>% select(gene, meta_analysis, h2_obs) %>% distinct()),
               by = c("gene", "meta_analysis"), suffix=c("_gwas_ctc", "_gene")) %>%
    mutate(gen_cov = rg * (sqrt(h2_obs_gene) * sqrt(h2_obs_gwas_ctc))) %>%
    filter(annot == "trans", p1 != p2)

p <- ggplot(combiend %>% filter(marker_gene, annot == "trans", p1 != p2), aes(h2_obs_gene, gen_cov, colour = p2)) +
    geom_point() +
    facet_wrap(vars(meta_analysis)) +
    theme(legend.position = "bottom")

ggsave("covariances_dotplot.pdf", p, width = 6, height = 6)


p <- ggplot(combiend %>% filter(marker_gene, annot == "trans", p1 != p2), aes(p2, gen_cov, colour = meta_analysis)) +
    geom_violin() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggsave("covariances_violin.pdf", p, width = 6, height = 6)


p <- ggplot(combiend %>% filter(annot == "trans", p1 != p2, meta_analysis == "20PSs"), aes(h2_obs_gene, gen_cov, colour = p2)) +
    geom_point() +
    facet_wrap(vars(marker_gene)) +
    theme(legend.position = "bottom")

ggsave("covariances_dotplot_mr.pdf", p, width = 6, height = 6)

p <- ggplot(combiend %>% filter(annot == "trans", p1 != p2, meta_analysis == "20PSs"), aes(p2, gen_cov, colour = marker_gene)) +
    geom_violin() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggsave("covariances_violin_mr.pdf", p, width = 6, height = 6)

```

```{r}

latent_factors_cis <- fread("latent_factors_cis_comp.txt") %>% mutate(annot="genome-wide")
latent_factors_trans <- fread("latent_factors_trans_comp.txt") %>% mutate(annot="trans")

merged_factors <- bind_rows(latent_factors_cis, latent_factors_trans) %>%
        filter(op == "=~") %>%
        mutate(marker_gene = rhs %in% mr_genes$ID)

summarised_cell_type_factors <- merged_factors %>% filter(lhs != "GeNonCtc") %>% group_by(rhs, annot) %>%
        summarise(sum_of_ct_explained_variance = sum(variance_explained))

summarised_sign_cell_type_factors <- merged_factors %>% filter(lhs != "GeNonCtc") %>% filter(p_value < 0.05) %>% group_by(rhs, annot) %>%
        summarise(sum_of_ct_sign_explained_variance = sum(variance_explained))

summarised_explained <- merged_factors %>%
        filter(lhs == "GeNonCtc", analysis=="all") %>%
        group_by(annot, marker_gene) %>%
        summarise(variance_explained_mean = mean(variance_explained))

ms4a1 <- merged_factors %>%
        filter(analysis=="all", rhs=="ENSG00000156738", annot=="genome-wide") %>%
        arrange(desc(variance_explained)) %>%
        mutate(cell_type = ordered(lhs, lhs))

p <- ggplot(ms4a1, aes(x=cell_type, y = variance_explained, fill = lhs == "GeNonCtc")) +
        geom_col(stat='identity') +
        coord_cartesian(ylim=c(0, 1.2)) +
        theme(axis.text.x = element_text(angle=45, vjust=0.5, hjust=1))

ggsave("explained_variance_ms4a1_h2_12-04_trans.pdf", p, width = 4, height = 4)


        filtered_pivot <- merged_factors %>%
        filter(lhs == "GeNonCtc", analysis=="all") %>%
        group_by(annot) %>%
        mutate(rank = base::rank(variance_explained)) %>%
        ungroup() %>%
        rename(gene_expression = variance_explained) %>%
        mutate(sum_of_cell_types = 1-gene_expression) %>%
        pivot_longer(
                cols=c("sum_of_cell_types", "gene_expression"),
                values_to = "variance_explained",
                names_to="type") %>%
        mutate(MS4A1 = rhs=="ENSG00000156738")


p <- ggplot(filtered_pivot, aes(x=rank, y=variance_explained, fill=type, alpha=p_value<0.05)) +
        geom_col(stat='identity', position = position_stack(reverse = TRUE)) +
        scale_fill_manual(values=c("sum_of_cell_types"="grey70", "gene_expression"="#D55E00"), labels=c("Cell-type\ncomposition", "Gene\nexpression"))+
        geom_point(data=. %>% filter(MS4A1)) +
        geom_hline(data=summarised_explained, mapping=aes(yintercept=variance_explained_mean), inherit.aes=F) +
        facet_wrap(vars(annot), nrow=2) +
        theme(legend.position = "bottom")

ggsave("explained_variance_bars_onesum_h2_12-04_trans.pdf", p, width = 5, height = 4)


filtered_pivot <- merged_factors %>%
        filter(lhs == "GeNonCtc", analysis=="all") %>%
        group_by(annot,marker_gene) %>%
        mutate(rank = base::rank(variance_explained)) %>%
        ungroup() %>%
        rename(gene_expression = variance_explained) %>%
        mutate(sum_of_cell_types = 1-gene_expression) %>%
        pivot_longer(
                cols=c("sum_of_cell_types", "gene_expression"),
                values_to = "variance_explained",
                names_to="type") %>%
        mutate(MS4A1 = rhs=="ENSG00000156738")


p <- ggplot(filtered_pivot, aes(x=rank, y=variance_explained, fill=type, alpha=p_value<0.05)) +
        geom_col(stat='identity', position = position_stack(reverse = TRUE)) +
        scale_fill_manual(values=c("sum_of_cell_types"="grey70", "gene_expression"="#D55E00"), labels=c("Cell-type\ncomposition", "Gene\nexpression"))+
        geom_point(data=. %>% filter(MS4A1)) +
        geom_hline(data=summarised_explained, mapping=aes(yintercept=variance_explained_mean), inherit.aes=F) +
        facet_wrap(vars(annot, marker_gene), nrow=2) +
        theme(legend.position = "bottom")

ggsave("explained_variance_bars_onesum_h2_12-04_marker.pdf", p, width = 5, height = 4)


filtered <- merged_factors %>%
        filter(lhs == "GeNonCtc", analysis=="all")


my_new_table <- tibble(h2_table_concat) %>% inner_join(filtered, by=c("gene_id"="rhs", "marker_gene", "annot"))

pivotted_table <- my_new_table %>%
        group_by(annot) %>%
        mutate(rank = base::rank(sum_of_variance_explained)) %>%
        pivot_longer(
             cols=c("sum_of_cell_types", "gene_expression"),
             values_to = "variance_explained",
             names_to="type")

p <- ggplot(pivotted_table, aes(x=rank, y=variance_explained, fill=type, alpha=p_value<0.05)) +
        geom_col(stat='identity', position = position_stack(reverse = TRUE)) +
        coord_cartesian(ylim= c(-1, 2)) +
        facet_wrap(vars(annot), nrow=2)
ggsave("explained_variance_bars_non_sign_h2_11-28_trans.pdf", p, width = 4, height = 4)

p <- ggplot(pivotted_table, aes(x=rank, y=variance_explained, fill=type, alpha=p_value<0.05)) +
        geom_col(stat='identity', position = position_dodge()) +
        coord_cartesian(ylim= c(-1, 2)) +
        facet_wrap(vars(annot), nrow=2)
ggsave("explained_variance_bars_dodge_h2_11-28_trans.pdf", p, width = 4, height = 4)

p <- ggplot(my_new_table, aes(h2_obs, variance_explained, colour = marker_gene, alpha = -log10(p_value))) +
        geom_point() + facet_wrap(vars(annot), nrow=2)
ggsave("explained_variance_ratio_h2_12-04_trans.pdf", p, width = 4, height = 4)

p <- ggplot(my_new_table, aes(x=marker_gene, y=ratio_variance_explained_ge, fill = -log10(p_value))) +
        geom_violin() + facet_wrap(vars(annot), nrow=2)
ggsave("explained_variance_ratio_violin_h2_11-28_trans.pdf", p, width = 4, height = 4)

p <- ggplot(filtered, aes(marker_gene, variance_explained, fill=p_value < 0.05)) + geom_violin() + facet_wrap(vars(annot), nrow=2)
ggsave("violin_explained_variance_h2_11-28_trans.pdf", p, width = 4, height = 4)

p <- ggplot(merged_factors)
```